### PR TITLE
Refactor service-ca and payload config controller (part 2)

### PR DESCRIPTION
**Pull request by mfojtik**
Follow-up for https://github.com/openshift/cluster-authentication-operator/pull/292

This continue to split the main operator loop. In this iteration the "block 2" is pushed into its own controller.

### CODE REVIEW

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "v4-0-config-system-service-ca",
+			Namespace: "openshift-authentication",
+			Labels: map[string]string{
+				"app": "oauth-openshift",
+			},
+			Annotations: map[string]string{
+				"service.alpha.openshift.io/inject-cabundle":          "true",
+				"service.alpha.openshift.io/serving-cert-secret-name": "v4-0-config-system-serving-cert",
```
Comment (stlaz) [inline pkg/controllers/serviceca/service_ca_controller.go:181]: this line does not belong here

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		configInformer.Config().V1().Authentications().Informer(),
+		configInformer.Config().V1().Ingresses().Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("ServiceCAController", recorder.WithComponentSuffix("service-ca-controller"))
+}
+
+func (c *serviceCAController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	// make sure we create the service before we start asking about service certs
+	_, _, err := resourceapply.ApplyService(c.services, syncCtx.Recorder(), getOAuthService())
```
Comment (stlaz) [inline pkg/controllers/serviceca/service_ca_controller.go:63]: this is a constant, you could move it to bindata now that we have them

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+	if _, _, err := v1helpers.UpdateStatus(c.operatorClient, updateConditionFuncs...); err != nil {
+		return err
+	}
+
+	// retry faster if we have degraded condition
+	// TODO: replace this with len(foundConditions) != 0 after all controllers are extracted.
+	if v1helpers.IsOperatorConditionTrue(foundConditions, "AuthServiceCAProgressing") {
+		return factory.SyntheticRequeueError
+	}
```
Comment (stlaz) [inline pkg/controllers/serviceca/service_ca_controller.go:118]: if you're to re-use this pattern, you should extract it somewhere

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 2 above)
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:63]: can this be a follow up? i want to minimize the diff.

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 3 above)
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:118]: already noticed that, again a follow up :-)

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 1 above)
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go]: fixed.

**Comment by mfojtik** [comment]
PullBuilderImageFailed: Failed pulling builder image.

/retest

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+)
+
+func init() {
+	utilruntime.Must(osinv1.Install(scheme))
+}
+
+// knownConditionNames lists all condition types used by this controller.
+// These conditions are operated and defaulted by this controller.
+// Any new condition used by this controller sync() loop should be listed here.
+var knownConditionNames = sets.NewString(
```
Comment (deads2k) [inline pkg/controllers/cliconfig/cli_config_controller.go:52]: you're going to all the trouble of splitting bits out and still have three conditions?  That seems weird.

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+		kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
+		operatorClient.Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("CLIConfig", recorder.WithComponentSuffix("cli-config-controller"))
+}
+
+func (c *cliConfigController) getRoute(ctx context.Context) (*routev1.Route, []operatorv1.OperatorCondition) {
+	// route is a pre-requirement for this sync
+	// it is created by metadata controller
+	// if route does not exists, do nothing and wait
+	route, err := c.route.Get(ctx, "oauth-openshift", metav1.GetOptions{})
```
Comment (deads2k) [inline pkg/controllers/cliconfig/cli_config_controller.go:89]: this stands out as not being a lister.

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "CLIConfigDegraded",
```
Comment (deads2k) [inline pkg/controllers/cliconfig/cli_config_controller.go:93]: This isn't related to the CLI.  You want this degraded message to point to where the customer should look for a fix or what the impact of the failure is.  In this case, IngressCreationDegraded?

Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go:93]: @deads2k the result of this controller sync loop is `"v4-0-config-system-cliconfig`

File: pkg/controllers/cliconfig/cli_config_controller.go (hunk 5 above)
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go:89]: sure, changed.

File: pkg/controllers/cliconfig/cli_config_controller.go (hunk 4 above)
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go:54]: now there is more than three :-)

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 3 above)
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:115]: (it is part of "part 3" ;-)

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 2 above)
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go]: nvmd. I already moved this to bindata here

File: pkg/controllers/cliconfig/cli_config_controller.go (hunk 6 above)
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go]: i mean i'm open for suggestions of the name for this controller :-)

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+
+func (c *cliConfigController) getService() (*corev1.Service, []operatorv1.OperatorCondition) {
+	// service is a pre-requirement for this sync
+	// it is created by serviceca controller
+	// if service does not exists, do nothing and wait
+	service, err := c.services.Services("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "OAuthServerServiceDegraded",
```
Comment (stlaz) [inline pkg/controllers/cliconfig/cli_config_controller.go:117]: suddenly you have two conditions that appear very similar: `OAuthServiceDegraded` and `OAuthServerServiceDegraded`. Why don't you keep one (preferrably at the controller that sets the service) and have the other controllers just log a failure/use a generic controller-specific condition as shown in the other comment?

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+
+func (c *cliConfigController) getRoute() (*routev1.Route, []operatorv1.OperatorCondition) {
+	// route is a pre-requirement for this sync
+	// it is created by metadata controller
+	// if route does not exists, do nothing and wait
+	route, err := c.route.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "OAuthServerIngressDegraded",
```
Comment (stlaz) [inline pkg/controllers/cliconfig/cli_config_controller.go:99]: Will every controller have its own "couldn't get route" condition now? Should it be something more generic, like `type: "CLIConfigControllerSyncDegraded"` which would cover most of these cases?

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+	}
+	return service, nil
+}
+
+func (c *cliConfigController) getAuthConfig(ctx context.Context) (*operatorv1.Authentication, []operatorv1.OperatorCondition) {
+	operatorConfig, err := c.auth.Authentications().Get(ctx, "cluster", metav1.GetOptions{})
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "OAuthConfigDegraded",
```
Comment (stlaz) [inline pkg/controllers/cliconfig/cli_config_controller.go:132]: similar to the other comments, does everyone get a similar condition type now?

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+			Annotations:     map[string]string{},
+			OwnerReferences: nil, // TODO
+		},
+		Data: map[string]string{
+			"v4-0-config-system-cliconfig": string(completeConfigBytes),
+		},
+	}
+}
+
+func getMasterCA() *string {
```
Comment (stlaz) [inline pkg/controllers/cliconfig/cli_config_controller.go:363]:
```golang
func sptr(s string) *string {
    return &s
}
```
and inline the string const above in this function to make this all more readable

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+			{
+				Type:    "OAuthConfigDegraded",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "ApplyFailed",
+				Message: fmt.Sprintf("Failed to apply CLI configuration %q: %v", expectedCLIConfig.Name, err),
+			},
+		}
+	}
+
+	return nil
```
Comment (stlaz) [inline pkg/controllers/cliconfig/cli_config_controller.go:323]: the actual logic gets covered up by construction of the conditions :/ I can try to deal with this in a follow-up

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+		configInformer.Config().V1().Authentications().Informer(),
+		configInformer.Config().V1().Ingresses().Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("ServiceCAController", recorder.WithComponentSuffix("service-ca-controller"))
+}
+
+func (c *serviceCAController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	// make sure we create the service before we start asking about service certs
+	if _, err := c.serviceLister.Services("openshift-authentication").Get("oauth-openshift"); err != nil {
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:62]: i would go Degraded in any error case here, service removed is one of them? do you think we need to handle 404 differently?

File: pkg/operator2/deployment.go [hunk 13]
```diff
@@ -88,6 +91,27 @@ func defaultDeployment(
 	return deployment, nil
 }
 
+// grabPrefixedConfig returns the configuration from the operator's observedConfig field
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/operator2/deployment.go:94]: maybe, i tried to separate late in dev cycle, I don't think it is critical to have the removals/additions right for the overall change, do you?

File: pkg/operator2/starter.go [hunk 14]
```diff
@@ -119,6 +120,8 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 			"oauth-openshift/cabundle.yaml",
 			"oauth-openshift/branding-secret.yaml",
 			"oauth-openshift/serviceaccount.yaml",
+			// oauth server:
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/operator2/starter.go:124]: i tried to separate assets that are related to oauth server vs. operator.

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+				"app": "oauth-openshift",
+			},
+			OwnerReferences: nil, // TODO
+		},
+	}
+}
+
+func (c *serviceCAController) getServiceCA(ctx context.Context, recorder events.Recorder) ([]operatorv1.OperatorCondition, error) {
+	cm := c.configMaps.ConfigMaps("openshift-authentication")
+	secret := c.secrets.Secrets("openshift-authentication")
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:137]: yes. will add it.

File: pkg/operator2/deployment.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/deployment.go:94]: I went commit by commit and this change seemed confusing, but if it's necessary for stuff to work, it's ok

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+		configMaps:     configMaps,
+		operatorClient: operatorClient,
+	}
+	return factory.New().WithInformers(
+		kubeInformersForTargetNamespace.Core().V1().Secrets().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().Services().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
+		configInformer.Config().V1().Authentications().Informer(),
+		configInformer.Config().V1().Ingresses().Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("ServiceCAController", recorder.WithComponentSuffix("service-ca-controller"))
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:55]: usually we add timed sync to every controller, for the case the informers get stuck

File: pkg/controllers/cliconfig/cli_config_controller.go [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+		operatorClient: operatorClient,
+		auth:           authentication,
+	}
+	return factory.New().WithInformers(
+		kubeInformersForTargetNamespace.Core().V1().Secrets().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().Services().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
+		routeInformer.Informer(),
+		operatorClient.Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("CLIConfig", recorder.WithComponentSuffix("cli-config-controller"))
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go:88]: same answer

File: pkg/operator2/starter.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/starter.go:124]: all of these assets are related to the oauth server, aren't they?

File: pkg/controllers/cliconfig/cli_config_controller.go (hunk 8 above)
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go:99]: if just one controller get stuck, then I would rather prefer each controller having its own condition.

File: pkg/controllers/serviceca/service_ca_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/serviceca/service_ca_controller.go:62]: right, I originally thought this controller attempted to sync the service itself as well, but that's done elsewhere.

File: pkg/operator2/starter.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/operator2/starter.go:124]: haven't checked :-) removed the comment.

File: pkg/controllers/serviceca/service_ca_controller.go [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+	if serviceCA.Annotations["service.alpha.openshift.io/inject-cabundle"] != "true" {
+		// return fmt.Errorf("config map missing injection annotation: %#v", ca)
+		// delete the service CA config map so that it is replaced with the proper one in next reconcile loop
+		opts := metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &serviceCA.UID}}
+		if err := cm.Delete(ctx, serviceCA.Name, opts); err != nil && !errors.IsNotFound(err) {
+			recorder.Warningf("InvalidServiceCAFailed", "Failed to delete invalid service CA config map %s: %v", spew.Sdump(serviceCA), err)
+		}
+		recorder.Eventf("InvalidServiceCA", "Deleted invalid service CA config map: %s", spew.Sdump(serviceCA))
+		return nil, factory.SyntheticRequeueError
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/controllers/serviceca/service_ca_controller.go:169]: added more context to event

File: pkg/controllers/cliconfig/cli_config_controller.go (hunk 7 above)
Comment (mfojtik) [reply pkg/controllers/cliconfig/cli_config_controller.go]: i went with OAuthConfigServiceDegraded

**Pull request merged**